src/compute/src/render/reduce.rs

```diff
@@ -947,37 +947,50 @@ where
         let arranged_output = input.scope().region_named("ReduceHierarchical", |inner| {
             let input = input.enter(inner);
 
-            // Gather the relevant values ordered by aggregation_index
-            let input = input.map(move |(key, row)| {
+            // Gather the relevant keys with their hashes along with values ordered by aggregation_index.
+            let mut stage = input.map(move |(key, row)| {
                 let binding = SharedRow::get();
                 let mut row_builder = binding.borrow_mut();
                 let mut row_packer = row_builder.packer();
                 let mut row_iter = row.iter();
                 for skip in skips.iter() {
                     row_packer.push(row_iter.nth(*skip).unwrap());
                 }
+                let values = row_builder.clone();
 
-                (key, row_builder.clone())
+                row_packer = row_builder.packer();
+                row_packer.push(Datum::UInt64(values.hashed()));
+                row_packer.extend(&key);
+                (row_builder.clone(), values)
             });
 
             // Repeatedly apply hierarchical reduction with a progressively coarser key.
-            let mut stage = input.map(move |(key, values)| ((key, values.hashed()), values));
             for b in buckets.into_iter() {
-                let input = stage.map(move |((key, hash), values)| ((key, hash % b), values));
+                let input = stage.map(move |(hash_key, values)| {
+                    let mut hash_key_iter = hash_key.iter();
+                    let hash = hash_key_iter.next().unwrap().unwrap_uint64() % b;
+
+                    let binding = SharedRow::get();
+                    let mut row_builder = binding.borrow_mut();
+                    let mut row_packer = row_builder.packer();
+                    row_packer.push(Datum::UInt64(hash));
+                    row_packer.extend(hash_key_iter);
+                    (row_builder.clone(), values)
+                });
 
                 // We only want the first stage to perform validation of whether invalid accumulations
                 // were observed in the input. Subsequently, we will either produce an error in the error
                 // stream or produce correct data in the output stream.
                 let negated_output = if err_output.is_none() {
                     let (oks, errs) = self
-                        .build_bucketed_negated_output::<_, Result<Row, (Row, u64)>>(
+                        .build_bucketed_negated_output::<_, Result<Row, Row>>(
                             &input,
                             aggr_funcs.clone(),
                         )
                         .map_fallible(
                             "Checked Invalid Accumulations",
                             |(key, result)| match result {
-                                Err((key, _)) => {
+                                Err(key) => {
```

Comment:
We will now print a `Row` here including the hash instead of simply a `Row` for the key. I argue that this is not of much consequence if you already have an invalid accumulation error to deal with.

Reply:
It can be of consequence if we have any of these errors in a materialized view. If the text changes, we cannot retract them anymore.

Reply:
Great catch! I changed this now to reconstruct a `Row` for the key only.

Reply:
Upon reflection, wouldn't the self-correcting scheme in `persist_sink` allow us to retract anyway? We'd read the old error via `persist_source`, retract it, and add the new error message.

Reply:
You're right! Please disregard my comment.